test: added markdown_link_check integration test

tomaswoj-eth opened the pull request:
### Description:
Added markdown_link_check for integration ci tests

### Checklist:
- [x]  markdown_link_check

Fixes part of #413

smoelius commented:
Hey, @tomaswoj-eth.

Sorry, what does this test do? (It looks like it does something other than what [`markdown-link-check`](https://github.com/tcort/markdown-link-check) does.)

tomaswoj-eth commented:
@smoelius Thanks for the feedback. I'm trying to determine the best approach for implementing this markdown link check test.

I see two main options:

1. **Call the npm package**
   - This would mirror the GitHub workflow exactly, using the `markdown-link-check` npm tool
   - Pros: Full HTTP validation as the original tool does
   - Cons: External dependency on npm/Node.js

2. **Pure Rust implementation**
   - Build a native solution using Rust's regex and optional HTTP client libraries
   - Pros: No external dependencies, consistent with the rest of the codebase
   - Cons: Would need to reimplement some of the npm tool's functionality

Which approach would you prefer? My goal is to meet the requirements from issue #413 while following the project's conventions. If you have a specific direction or implementation preference, I'd appreciate your guidance.

smoelius commented:
I am content to use the `npm` tool.

I agree that relying on `npm` is kind of ugly, but it installs quickly and does the job, so I think it is fine.

Note that the current use of `markdown-link-check` checks only README.md (which I also think is fine). The Dylint version checks all markdown files in the repository.

Here is a simpler version that checks only README.md: https://github.com/trailofbits/cast_checks/blob/61c7e7e3be7a212b05a5d36e7483985145ecdb37/cast_checks/tests/ci.rs#L59-L76

tomaswoj-eth commented:
@smoelius I've tried to implement the code from https://github.com/trailofbits/cast_checks/blob/61c7e7e3be7a212b05a5d36e7483985145ecdb37/cast_checks/tests/ci.rs#L59-L76 as you suggested, had to make some adjustments, but I think I got it working. Check if the changes are what you expected, if any more adjustments need to be made, please let me know.

tomaswoj-eth commented:
any reviews @smoelius ?

smoelius commented:
> any reviews @smoelius ?

Please forgive me for saying this, but it's been just over an hour since your update. :)

I'll try to get to this later on today.

I do appreciate you working on this, though.

smoelius commented:
> @smoelius I've tried to implement the code from https://github.com/trailofbits/cast_checks/blob/61c7e7e3be7a212b05a5d36e7483985145ecdb37/cast_checks/tests/ci.rs#L59-L76 as you suggested, had to make some adjustments, but I think I got it working. Check if the changes are what you expected, if any more adjustments need to be made, please let me know.

I want to reiterate that I appreciate you working on this.

But I am a little confused. After copying the test from `cast_checks`, I did have to change `assert` to `logged_assert`. But after that, everything worked fine.

What necessitated the changes besides `assert` -> `logged_assert`?

tomaswoj-eth commented:
@smoelius I added those for local checking to see if the implementation worked. I've removed them in the current commit also I think changes to the ci.yml workflow file is required, I did not do those as I thought it is something you might want to do for security. Please go over my changes when you're free. I think this is ready to be merged. Nits are welcomed.

smoelius commented:
Thanks, @tomaswoj-eth. 🙂
